### PR TITLE
Updates to error handling

### LINKED ISSUE #466: Prepare error handling for MUSICA library
In preparation for implementing a solution for error handling in the MUSICA library, ensure that MICM consistently throws when errors are encountered, and use custom exceptions that can be caught and sent through the MUSICA C API in some way.

**Acceptance Criteria**
- All errors encountered in MICM throw custom exceptions

**Ideas**
- Explore `std::error_code` and `std::error_condition`

### PR DESCRIPTION
**Pull request by mattldawson**
Adds custom error codes to MICM classes and updates existing error handling to use them. The hope is that in the MUSICA library we can catch these errors and turn them into meaningful codes/messages that can be returned through the C API.

closes #466

### CODE REVIEW

**Review by K20shores — COMMENTED** [review]
I like it! I guess one concern is that we are potentially masking system errors in our tests if they were ever to be thrown and happen to have the same numerical value as the micm error values. 

And on that note, in MUSICA, I don't see how we will be able to identify exactly which error was thrown since we have overlapping codes. It seems like we would need to know exactly which part of the code threw this so that we know which `enum class` to match the error code to. What am I missing?

**Comment by mattldawson** [comment]
> I like it! I guess one concern is that we are potentially masking system errors in our tests if they were ever to be thrown and happen to have the same numerical value as the micm error values.
> 
> And on that note, in MUSICA, I don't see how we will be able to identify exactly which error was thrown since we have overlapping codes. It seems like we would need to know exactly which part of the code threw this so that we know which `enum class` to match the error code to. What am I missing?

Yeah, I was starting to work through that in the MUSICA library. One option might be to use a combination of the category name and the `int` value of the enum. Something like:
```
struct Error {
  int code_;
  const char* category_;
  const char* message_;
};
```
This way on the user side (even in Fortran, and I'm guessing Python), you could use `strcmp(error.category_, "MICM Matrix") == 0 && error.code_ == 2` to identify our `MicmMatrixErrc::InvalidVector` error. And this should avoid us masking other system errors. What do you think?

**Comment by WardF** [comment]
I'm taking a look at this today; getting caught up on the conversation. In C, software error codes are typically negative, and system error codes are typically positive, to avoid collisions/confusion.  I'm about to review the PR changes so perhaps it's already in place, but in the netCDF project we have an `errors.h` file which we use to define all the errors in one place and easier developer reference.

**Comment by K20shores** [comment]
> I'm taking a look at this today; getting caught up on the conversation. In C, software error codes are typically negative, and system error codes are typically positive, to avoid collisions/confusion. I'm about to review the PR changes so perhaps it's already in place, but in the netCDF project we have an `errors.h` file which we use to define all the errors in one place and easier developer reference.

I am a fan of having all of the errors in one place as well. It may mean a long file but it also means we don't need to have a specific error class for each header in micm. @mattldawson would you have anything against a single error file?

**Comment by sjsprecious** [comment]
For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?

**Comment by mwaxmonsky** [comment]
> For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?

If we do, we might be able to leverage the C++ library: https://en.cppreference.com/w/cpp/utility/source_location

**Comment by mattldawson** [comment]
> For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?

My thinking is that the errors we return to API users should be clear enough for them to act upon without having to look through our source code. (Problems with configuration files, passing improperly sized arrays to API functions, etc.) If we're just interested in bug checking our code at runtime, we can do this with asserts instead of throwing exceptions. What does everyone think?

**Review by WardF — APPROVED** [review]
The functionality of this looks good, albeit I'm taking a 10,000 foot view.  My one thought is that the code would benefit from some comments; while a lot of it is self evident now, it might be less so down the road.  "Code needing more in-line comments" is hardly novel feedback, I acknowledge XD.  Particularly since it could apply to any number of my own PRs.  But since I'm looking at this as an outsider and not a core developer, the function is less immediately obvious to me, and the lack of comments stands out more.

**Comment by mattldawson** [comment]
> > I'm taking a look at this today; getting caught up on the conversation. In C, software error codes are typically negative, and system error codes are typically positive, to avoid collisions/confusion. I'm about to review the PR changes so perhaps it's already in place, but in the netCDF project we have an `errors.h` file which we use to define all the errors in one place and easier developer reference.
> 
> I am a fan of having all of the errors in one place as well. It may mean a long file but it also means we don't need to have a specific error class for each header in micm. @mattldawson would you have anything against a single error file?

I like that too. The only thing that's a little unclear to me is how the dependencies would work. If the error codes live in MUSICA, would MICM and TUV-x depend on MUSICA? Or if they live in MICM and TUV-x, would MICM and TUV-x depend on each other?

**Comment by WardF** [comment]
> > > I'm taking a look at this today; getting caught up on the conversation. In C, software error codes are typically negative, and system error codes are typically positive, to avoid collisions/confusion. I'm about to review the PR changes so perhaps it's already in place, but in the netCDF project we have an `errors.h` file which we use to define all the errors in one place and easier developer reference.
> > 
> > 
> > I am a fan of having all of the errors in one place as well. It may mean a long file but it also means we don't need to have a specific error class for each header in micm. @mattldawson would you have anything against a single error file?
> 
> I like that too. The only thing that's a little unclear to me is how the dependencies would work. If the error codes live in MUSICA, would MICM and TUV-x depend on MUSICA? Or if they live in MICM and TUV-x, would MICM and TUV-x depend on each other?

If each project maintained a list of errors, their own `errors.h`, or `<project>_errors.h` which could be included as part of error handling, would that avoid explicit dependencies between the projects for error handling? I feel like that's a naive question, but it's the first that springs to mind.

**Comment by WardF** [comment]
> > For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?
> 
> My thinking is that the errors we return to API users should be clear enough for them to act upon without having to look through our source code. (Problems with configuration files, passing improperly sized arrays to API functions, etc.) If we're just interested in bug checking our code at runtime, we can do this with asserts instead of throwing exceptions. What does everyone think?

From experience, `ASSERT` is great when it works for the intended purpose, but when it leaks out into user-space (e.g. unexpected/unanticipated behavior results in an 'unreachable' `ASSERT` being reached by a user at runtime) it can become a bit of a pain. So I guess with careful planning, using an `ASSERT` for debugging purposes instead of FILE and LINE would work.  +100 to the idea of 'clear error messaging'.  

From a user support perspective, I wonder now about a user sending you details around a problem they had.  Would the error be enough for you to debug what's going on? Or would `__FILE__` and `__LINE__` be useful information to you then?  Just thinking out loud, clearly.

**Comment by mattldawson** [comment]
> > > > I'm taking a look at this today; getting caught up on the conversation. In C, software error codes are typically negative, and system error codes are typically positive, to avoid collisions/confusion. I'm about to review the PR changes so perhaps it's already in place, but in the netCDF project we have an `errors.h` file which we use to define all the errors in one place and easier developer reference.
> > > 
> > > 
> > > I am a fan of having all of the errors in one place as well. It may mean a long file but it also means we don't need to have a specific error class for each header in micm. @mattldawson would you have anything against a single error file?
> > 
> > 
> > I like that too. The only thing that's a little unclear to me is how the dependencies would work. If the error codes live in MUSICA, would MICM and TUV-x depend on MUSICA? Or if they live in MICM and TUV-x, would MICM and TUV-x depend on each other?
> 
> If each project maintained a list of errors, their own `errors.h`, or `<project>_errors.h` which could be included as part of error handling, would that avoid explicit dependencies between the projects for error handling? I feel like that's a naive question, but it's the first that springs to mind.

I like that idea. Would you expect the error header to just define the `int` codes and the category strings? Something like:
```
#define ERR_CAT_MICM_MATRIX "MICM Matrix"
#define ERR_CODE_MICM_ROW_SIZE_MISMATCH 1
#define ERR_CODE_MICM_INVALID_VECTOR 2
```
and then update, e.g., the Matrix class to use them:
```
enum class MicmMatrixErrc
{
  RowSizeMismatch = ERR_CODE_MICM_ROW_SIZE_MISMATCH,
  InvalidVector = ERR_CODE_MICM_INVALID_VECTOR,
};

namespace
{
  class MicmMatrixErrorCategory : public std::error_category
  {
   public:
    const char *name() const noexcept override
    {
      return ERR_CAT_MICM_MATRIX;
    }
    ...
```
The only thing might be that we couldn't use the same label for specific errors in different categories. (like if INVALID_VECTOR is an error in another class like SparseMatrix)

**Comment by mattldawson** [comment]
> The functionality of this looks good, albeit I'm taking a 10,000 foot view. My one thought is that the code would benefit from some comments; while a lot of it is self evident now, it might be less so down the road. "Code needing more in-line comments" is hardly novel feedback, I acknowledge XD. Particularly since it could apply to any number of my own PRs. But since I'm looking at this as an outsider and not a core developer, the function is less immediately obvious to me, and the lack of comments stands out more.

I have gone back and forth over time between 'comment everything' and 'try to write self-documenting code', and I don't know where I'm at right now. I'm happy to add more comments though. Was there anything in particular that you noticed should have more documentation?

**Comment by mattldawson** [comment]
Another benefit to the single file for errors might be that if we just use CPP definitions, we can include the same header in the fortran wrapper code and maybe cut down on the passing back and forth of strings between c and fortran or keeping this list in two places.

**Comment by mattldawson** [comment]
> > > For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?
> > 
> > 
> > My thinking is that the errors we return to API users should be clear enough for them to act upon without having to look through our source code. (Problems with configuration files, passing improperly sized arrays to API functions, etc.) If we're just interested in bug checking our code at runtime, we can do this with asserts instead of throwing exceptions. What does everyone think?
> 
> From experience, `ASSERT` is great when it works for the intended purpose, but when it leaks out into user-space (e.g. unexpected/unanticipated behavior results in an 'unreachable' `ASSERT` being reached by a user at runtime) it can become a bit of a pain. So I guess with careful planning, using an `ASSERT` for debugging purposes instead of FILE and LINE would work. +100 to the idea of 'clear error messaging'.
> 
> From a user support perspective, I wonder now about a user sending you details around a problem they had. Would the error be enough for you to debug what's going on? Or would `__FILE__` and `__LINE__` be useful information to you then? Just thinking out loud, clearly.

In going through the process of making these changes, I think it would be pretty clear what part of the code threw the error (the error categories are specific to one class, or in one case two classes that extend the same base class, and the individual errors are mostly only thrown in one place, or a few places at most).

To me what's important is that these errors always indicate a problem that the user should address, not one that indicates a bug in our code. If we also want to make it easier for users to report bugs in our code, what if we added something like an `InternalMicmErrc`? If we add some run-time bug check to our code that trips, it could throw this and include the file and line number of the detailed message. The user would get something like "Internal MICM Error(file, line): Please file a bug report..." What does everyone think of something like this?

**Comment by sjsprecious** [comment]
> > > > For the error handling, would it be better to add the `__FILE__` and `__LINE__` information so that a user can be directed to the exact location where the error occurs?
> > > 
> > > 
> > > My thinking is that the errors we return to API users should be clear enough for them to act upon without having to look through our source code. (Problems with configuration files, passing improperly sized arrays to API functions, etc.) If we're just interested in bug checking our code at runtime, we can do this with asserts instead of throwing exceptions. What does everyone think?
> > 
> > 
> > From experience, `ASSERT` is great when it works for the intended purpose, but when it leaks out into user-space (e.g. unexpected/unanticipated behavior results in an 'unreachable' `ASSERT` being reached by a user at runtime) it can become a bit of a pain. So I guess with careful planning, using an `ASSERT` for debugging purposes instead of FILE and LINE would work. +100 to the idea of 'clear error messaging'.
> > From a user support perspective, I wonder now about a user sending you details around a problem they had. Would the error be enough for you to debug what's going on? Or would `__FILE__` and `__LINE__` be useful information to you then? Just thinking out loud, clearly.
> 
> In going through the process of making these changes, I think it would be pretty clear what part of the code threw the error (the error categories are specific to one class, or in one case two classes that extend the same base class, and the individual errors are mostly only thrown in one place, or a few places at most).
> 
> To me what's important is that these errors always indicate a problem that the user should address, not one that indicates a bug in our code. If we also want to make it easier for users to report bugs in our code, what if we added something like an `InternalMicmErrc`? If we add some run-time bug check to our code that trips, it could throw this and include the file and line number of the detailed message. The user would get something like "Internal MICM Error(file, line): Please file a bug report..." What does everyone think of something like this?

Ah, I see. I am thinking about the code bug and I want the error message to be direct and pointed to the specific source code for faster debugging later. But regarding a user error, what you and others propose here makes sense to me, and using an error header file with enumed error code sounds a good idea. Thanks for your clarification.

**Comment by boulderdaze** [comment]
> > I like it! I guess one concern is that we are potentially masking system errors in our tests if they were ever to be thrown and happen to have the same numerical value as the micm error values.
> > And on that note, in MUSICA, I don't see how we will be able to identify exactly which error was thrown since we have overlapping codes. It seems like we would need to know exactly which part of the code threw this so that we know which `enum class` to match the error code to. What am I missing?
> 
> Yeah, I was starting to work through that in the MUSICA library. One option might be to use a combination of the category name and the `int` value of the enum. Something like:
> 
> ```
> struct Error {
>   int code_;
>   const char* category_;
>   const char* message_;
> };
> ```
> 
> With this to turn `std::system_error` into an `Error`:
> 
> ```
> Error ToError(const std::system_error& e)
> {
>     Error error;
>     error.code_ = e.code().value();
>     error.category_ = e.code().category().name();
>     error.message_ = e.what();
>     return error;
> }
> ```
> 
> This way on the user side (even in Fortran, and I'm guessing Python), you could use `strcmp(error.category_, "MICM Matrix") == 0 && error.code_ == 2` to identify our `MicmMatrixErrc::InvalidVector` error. And this should avoid us masking other system errors. What do you think?

I like this idea

**Review by boulderdaze — APPROVED** [review]
Looks good! Many interesting discussions above. I am not very familiar with the `system_error` header but we might be able to create a macro to avoid having to create each error category class to the applicable classes? something like, 
`MICM_ERROR(category, error code, message)`
Just a thought, thanks for the work! I think it's quite tricky to design the error handler since we also have to consider MUSICA.

**Comment by mattldawson** [comment]
Based on the discussion, I'll go ahead and create a micm error header and add an internal error category that includes the filename and line number in the message. (I'm not sure if we actually have any run-time checks that I can apply this to, but if I can't find one I'll just add a test that throws this error.)

I'll request a re-review once that's ready, but let me know if you would prefer not to review this again and I won't include you in the re-review.

Thanks for the discussion everyone!

File: include/micm/configure/solver_config.hpp
Comment (K20shores) [inline include/micm/configure/solver_config.hpp:31]: Now that we have a better error system, should we think about removing the success status since it's not an error? Rather than checking the return value throughout the configuration, we can throw an error when we arrive at one

File: include/micm/util/error.hpp
Comment (K20shores) [inline include/micm/util/error.hpp:54]: I like the categories, but I still worry that we could end up in a situation (not sure how), that we only have the integer value. What do #defines gain us over `static const strings` and one large `enum class` (for which we don't manually define the values)?

File: include/micm/util/error.hpp
Comment (K20shores) [inline include/micm/util/error.hpp:7]: why this instead of #pragma once?

**Review by K20shores — COMMENTED** [review]
Really like this. I left other comments but none of them would prevent this from being merged

**Review by K20shores — APPROVED** [review]

File: include/micm/configure/solver_config.hpp
Comment (mattldawson) [reply include/micm/configure/solver_config.hpp:31]: Good point. I'll do that now.

File: include/micm/util/error.hpp
Comment (mattldawson) [reply include/micm/util/error.hpp:7]: If we include this in a fortran file, any comments would have to be flagged with `!` instead of `//`. This was just to get the CPP to remove these lines entirely

File: include/micm/solver/jit_linear_solver.inl
Comment (mwaxmonsky) [inline include/micm/solver/jit_linear_solver.inl:44]: Might be worth having a message why it's invalid?

File: include/micm/solver/jit_lu_decomposition.inl
Comment (mwaxmonsky) [inline include/micm/solver/jit_lu_decomposition.inl:38]: Might be worth having a message why it's invalid?

File: include/micm/util/error.hpp
Comment (mattldawson) [reply include/micm/util/error.hpp:54]: I could go either way. There is something about having one big bucket for things that makes me nervous, but I can't actually think of a real example of how this would be a problem in this case.

The one argument for `#defines` is that  it allows this file to be included in fortran files as well, and might avoid some duplication or difficult passing of strings, but this isn't a super big deal.

I guess, my suggestion would be to keep the categories until we find ourselves in a situation where we only have the integer code for some reason and decide then if we should reorganize. What do you think?

File: include/micm/solver/jit_rosenbrock.hpp
Comment (mwaxmonsky) [inline include/micm/solver/jit_rosenbrock.hpp:90]: Might be worth having a message why it's invalid?

File: include/micm/solver/jit_rosenbrock.hpp
Comment (mwaxmonsky) [inline include/micm/solver/jit_rosenbrock.hpp:116]:
```suggestion
        throw std::system_error(make_error_code(MicmJitErrc::MissingJitFunction), std::source_location::current().function_name());
```

File: include/micm/util/matrix.hpp
Comment (mwaxmonsky) [inline include/micm/util/matrix.hpp:55]: Might be worth having a message why it's mismatched for easier debugging?

File: include/micm/util/matrix.hpp
Comment (mwaxmonsky) [inline include/micm/util/matrix.hpp:168]: Might be worth having a message why it's invalid?

File: include/micm/util/vector_matrix.hpp
Comment (mwaxmonsky) [inline include/micm/util/vector_matrix.hpp:103]: Might be worth having a message why it's mismatched for easier debugging?

File: include/micm/util/vector_matrix.hpp
Comment (mwaxmonsky) [inline include/micm/util/vector_matrix.hpp:221]: Might be worth having a message why it's invalid?

File: src/solver/rosenbrock.cu
Comment (mwaxmonsky) [inline src/solver/rosenbrock.cu:291]: Might be worth having a message why it's mismatched for easier debugging?

File: src/solver/rosenbrock.cu
Comment (mwaxmonsky) [inline src/solver/rosenbrock.cu:307]:
```suggestion
          throw std::system_error(make_error_code(MicmCudaRosenbrockSolverErrc::CublasError), "cublasStatus_t came back as " + stat);
```
Just to provide the actual error code to the user.

**Review by mwaxmonsky — APPROVED** [review]
Looks great @mattldawson!

Nothing worth holding this up but we might want to expand on some of the error messages if it makes sense and use the std library to get the function name in one place.

File: include/micm/solver/jit_rosenbrock.hpp
Comment (K20shores) [reply include/micm/solver/jit_rosenbrock.hpp:116]: that's a neat trick

File: src/solver/rosenbrock.cu
Comment (K20shores) [reply src/solver/rosenbrock.cu:307]: if that's the case should the printf the line above be removed?

Comment (mwaxmonsky) [reply src/solver/rosenbrock.cu:307]: I would say yes since I don't think we need to print to stdout anymore if there's an error since we are throwing the exception now.

File: include/micm/solver/jit_rosenbrock.hpp
Comment (mwaxmonsky) [reply include/micm/solver/jit_rosenbrock.hpp:116]: Oh, this requires the `#include <source_location>` header.

File: include/micm/configure/solver_config.hpp
Comment (mattldawson) [reply include/micm/configure/solver_config.hpp]: done!

File: include/micm/solver/jit_lu_decomposition.inl
Comment (mattldawson) [reply include/micm/solver/jit_lu_decomposition.inl]: done!

File: include/micm/solver/jit_rosenbrock.hpp
Comment (mattldawson) [reply include/micm/solver/jit_rosenbrock.hpp]: done!

File: include/micm/solver/jit_rosenbrock.hpp
Comment (mattldawson) [reply include/micm/solver/jit_rosenbrock.hpp]: pretty cool! done!

File: include/micm/util/matrix.hpp
Comment (mattldawson) [reply include/micm/util/matrix.hpp]: done!

File: include/micm/util/matrix.hpp
Comment (mattldawson) [reply include/micm/util/matrix.hpp:170]: done!

File: include/micm/util/vector_matrix.hpp
Comment (mattldawson) [reply include/micm/util/vector_matrix.hpp]: done!

File: include/micm/util/vector_matrix.hpp
Comment (mattldawson) [reply include/micm/util/vector_matrix.hpp]: done!

File: src/solver/rosenbrock.cu
Comment (mattldawson) [reply src/solver/rosenbrock.cu]: I ended up switching this to use the new `INTERNAL_ERROR()` because it looked to me more like a check for a bug in our code. But, if that's not correct, I can change it back to throw one of our user errors

File: src/solver/rosenbrock.cu
Comment (mattldawson) [reply src/solver/rosenbrock.cu]: yeah, I agree - I try to remove output to stdout and stderr when I find them. This should be updated now.

**Review by mwaxmonsky — APPROVED** [review]
Thanks for addressing those comments @mattldawson!

**Comment by WardF** [comment]
Reviewing now as well, although I'm sure I'll be the slowest due to relative unfamiliarity; I don't anticipate finding any issues, but I'm trying to digest what I'm looking at instead of skimming it :)

File: include/micm/util/error.hpp
Comment (boulderdaze) [reply include/micm/util/error.hpp:7]: How about not including copyright for this specific file? I think that makes sense that we don't want to duplicate the error file for fortran but it could also be confusing for MICM being standalone CPP project to have this unusual usage of macro to cater for fortran. What do you think?

**Review by WardF — APPROVED** [review]

**Review by boulderdaze — APPROVED** [review]
Looks great! Thank you for the work! I have one suggestion but it shouldn't prevent this branch from being merged.

**Comment by mattldawson** [comment]
I agree about not making things confusing just to cater to fortran, but I think we should try to have the copyright and license on each source code file. What if we revisit this after the fortran wrapper is updated? It could be the case that we don't even need it in fortran and can just use normal comments.

**Comment by boulderdaze** [comment]
> > I agree about not making things confusing just to cater to fortran, but I think we should try to have the copyright and license on each source code file. What if we revisit this after the fortran wrapper is updated? It could be the case that we don't even need it in fortran and can just use normal comments.
> 
> (I added this to our post-development review issue #211)

Sounds good, thank you!

**Pull request merged**
